usaspending_api/etl/management/commands/combine_universal_transaction_matview_chunks.py

```diff
@@ -0,0 +1,200 @@
+        chunk_count = options["chunk_count"]
+
+        logger.info(f"Chunk Count: {chunk_count}")
+
+        with Timer("Recreating table"):
+            self.recreate_matview()
+
+        with Timer("Inserting data into table"):
+            self.insert_matview_data(chunk_count)
+
+        with Timer("Creating table indexes"):
+            self.create_indexes()
+
+        with Timer("Swapping Tables/Indexes"):
+            self.swap_matviews()
+
+        if not options["keep_old_data"]:
+            self.remove_old_data(chunk_count)
+
+        if options["analyze"]:
+            with Timer("Analyzing Table"):
+                self.analyze_matview()
+
+        with Timer("Granting Table Permissions"):
+            self.grant_matview_permissions()
+
+    def recreate_matview(self):
+        with connection.cursor() as cursor:
+            # This table was previously a Matview, so the DROP may fail
+            try:
+                cursor.execute(RECREATE_TABLE_SQL.format(old_object_type="TABLE"))
+            except Exception as e:
+                if "is not a table" in str(e):
+                    logger.info(
+                        "universal_transaction_matview_temp existed, but not as table. Trying to drop as Matview"
+                    )
+                    cursor.execute(RECREATE_TABLE_SQL.format(old_object_type="MATERIALIZED VIEW"))
+
+    def insert_matview_data(self, chunk_count):
+        loop = asyncio.new_event_loop()
+        tasks = []
+
+        for current_chunk in range(0, chunk_count):
+            tasks.append(
+                asyncio.ensure_future(
+                    async_run_creates(
+                        INSERT_INTO_TABLE_SQL.format(current_chunk=current_chunk),
+                        wrapper=Timer(f"Insert into table {current_chunk}"),
+                    ),
+                    loop=loop,
+                )
+            )
+
+        loop.run_until_complete(asyncio.gather(*tasks))
+        loop.close()
+
+    def create_indexes(self):
+        loop = asyncio.new_event_loop()
+        tasks = []
+        i = 0
+        for sql in sqlparse.split(TABLE_INDEX_SQL):
+            tasks.append(
+                asyncio.ensure_future(async_run_creates(sql, wrapper=Timer(f"Creating Index {i}"),), loop=loop,)
+            )
+            i += 1
+
+        loop.run_until_complete(asyncio.gather(*tasks))
+        loop.close()
+
+    def swap_matviews(self):
+        with connection.cursor() as cursor:
+            # This table was previously a Matview, so the DROP may fail
+            try:
+                cursor.execute(SWAP_TABLES_SQL.format(old_object_type="TABLE"))
+            except Exception as e:
+                if "is not a table" in str(e):
+                    logger.info(
+                        "universal_transaction_matview_temp existed, but not as table. Trying to drop as Matview"
+                    )
+                    cursor.execute(SWAP_TABLES_SQL.format(old_object_type="MATERIALIZED VIEW"))
```

Comment:
This batch of SQL statements in `SWAP_TABLES_SQL` needs to be run in a single atomic database transaction. 

- It all either needs to succeed
- Or it all needs to fail

For instance, done this way, there will be a period of time where `universal_transaction_matview` doesn't exist at all. Between the rename to `_old` and the rename from `_temp` while the indexes are being altered. 

And then the indexes are coming online for the new one 1 index at a time.

Since this is a management command, the `connection` is a django managed connection, so you _could_ using the 

```python
from django.db import transaction

@transaction.atomic
def swap_matviews(self):
```

See: https://docs.djangoproject.com/en/2.2/topics/db/transactions/#controlling-transactions-explicitly

Reply:
This has been added